extract_sets/training_sets.py

```diff
@@ -0,0 +1,398 @@
+            return img_path
+    return None
+
+
+def _determine_namespace(xml_data):
+    root_tag = xml_data.xpath('namespace-uri(.)')
+    return [k for (k, v) in XML_NS.items() if v == root_tag][0]
+
+
+class TrainingSets:
+    """
+    Set of textlines and corresponding image parts created from given
+    source if text_len > min
+    """
+
+    def __init__(self, path_xml_data, path_image_data):
+        self.xdpi = None
+        self.ydpi = None
+        self.path_out = None
+        (self.set_label, _) = os.path.splitext(os.path.basename(path_xml_data))
+        self.xml_data = etree.parse(path_xml_data).getroot()
+        self.path_image_data = path_image_data
+        if not self.path_image_data:
+            self._resolve_image_path(path_xml_data)
+        self.image_data = TrainingSets._load_image(self.path_image_data)
+        self._read_dpi()
+
+    def _resolve_image_path(self, path_xml_data):
+        self.path_image_data = resolve_image_path(path_xml_data)
+
+    @staticmethod
+    def _load_image(path_image_data):
+        return cv2.imread(path_image_data, cv2.IMREAD_GRAYSCALE)
+
+    def write_data(self, text_line: TextLine,
+                   image_handle, path_out, prefix):
+        """Serialize training data pairs"""
+
+        if not path_out:
+            path_out = prefix + self.set_label
+        self.path_out = path_out
+        os.makedirs(path_out, exist_ok=True)
+
+        content = text_line.get_textline_content()
+        if content:
+            file_name = self.set_label + '_' + text_line.element_id + '.gt.txt'
+            file_path = os.path.join(path_out, file_name)
+            with open(file_path, 'w', encoding="utf8") as fhdl:
+                fhdl.write(content)
+
+            img_frame = TrainingSets._extract_frame(
+                image_handle, text_line)
+            file_name = self.set_label + '_' + text_line.element_id + '.tif'
+            file_path = os.path.join(path_out, file_name)
+
+            if img_frame.any():
+                params = self._calculate_tiff_param()
+                if params:
+                    cv2.imwrite(file_path, img_frame, params)
+                else:
+                    cv2.imwrite(file_path, img_frame)
+
+    def write_all(self, training_datas):
+        """Serialize training data pairs"""
+
+        contents = [d.get_textline_content() + '\n' for d in training_datas]
+        file_name = self.set_label + SUMMARY_SUFFIX
+        file_path = os.path.join(self.path_out, file_name)
+        with open(file_path, 'w', encoding="utf8") as fhdl:
+            fhdl.writelines(contents)
+
+    @staticmethod
+    def _extract_frame(image_handle, text_line):
+        the_shape = text_line.shape
+        the_box = the_shape.bounds
+        start_vpos = int(the_box[1])
+        end_vpos = int(the_box[3])
+        start_hpos = int(the_box[0])
+        end_hpos = int(the_box[2])
+        return image_handle[start_vpos:end_vpos, start_hpos:end_hpos]
```

Comment:
This unnecessarily throws away any polygon segmentation and just uses the bounding boxes for cropping. This could be problematic if there are overlaps between bounding boxes (as in densely typeset historic texts, or coarse line segmentation). Of course, tesstrain could still be run with PSM=7 or such, but why not use the polygon for masking? ALTO v4 has `Shape/Polygon/POINTS`, and PAGE of course is always polygonal.

The only problem with masking could be the background colour: If the image is not binarized when imreading, simly using white could create an artificial "edge". But you can easily use PIL.ImageStat.Stat's or Numpy's `median` here. See `ocrd_utils.image_from_polygon` for a full solution.

Reply:
Also, it could be useful to look for an `AlternativeImage` on the line level instead of cropping here.

Reply:
You're right, I'm still not finished to work this out.

Reply:
> fd09bb0

Wow, that was super-fast! Will take me some time to re-review. Have never seen use of `cv2.fillConvexPoly` before, will try it out. Creating background with (blurred) random noise is a very interesting idea (but I'm afraid this can currently overflow and wrap-around).